### PR TITLE
feat: add guards for absent registry credentials

### PR DESCRIPTION
**Pull request by innovaker**
Slightly different approach than #12.  Please check commit message.

### CODE REVIEW

**Comment by innovaker** [comment]
This PR is built on the HEAD of #23.

**Comment by innovaker** [comment]
### Testing
- without `DOCKER_HUB_USERNAME`
  - Candidate: https://github.com/innovaker/zmk-docker/runs/2553975788
  - Release: https://github.com/innovaker/zmk-docker/runs/2553989468

- without `GHCR_USERNAME`
  - Candidate: https://github.com/innovaker/zmk-docker/runs/2554000915
  - Release: https://github.com/innovaker/zmk-docker/actions/runs/831095980

- without `DOCKER_HUB_TOKEN`
  - Candidate: https://github.com/innovaker/zmk-docker/actions/runs/831099834
  - Release: https://github.com/innovaker/zmk-docker/actions/runs/831104333

- without `GHCR_TOKEN`
  - Candidate: https://github.com/innovaker/zmk-docker/runs/2554056979
  - Release: https://github.com/innovaker/zmk-docker/actions/runs/831115193

- with everything again (check new tokens):
  - Candidate: https://github.com/innovaker/zmk-docker/actions/runs/831127833
  - Release: https://github.com/innovaker/zmk-docker/actions/runs/831131064

File: .github/workflows/containers.yml [hunk 1]
```diff
@@ -22,20 +22,31 @@ jobs:
...
+          echo '::set-output name=ghcr::${{ toJSON(secrets.GHCR_USERNAME != null && secrets.GHCR_TOKEN != null) }}'
       - name: Login to GitHub Container Registry
+        if: ${{ fromJSON(steps.credentials.outputs.ghcr) || startsWith(github.ref, 'refs/tags') }}
         uses: docker/login-action@v1
         with:
           registry: ghcr.io
           username: ${{ secrets.GHCR_USERNAME }}
           password: ${{ secrets.GHCR_TOKEN }}
       - name: Login to Docker Hub
+        if: ${{ fromJSON(steps.credentials.outputs.docker-hub) || startsWith(github.ref, 'refs/tags') }}
```
Comment (petejohanson) [inline .github/workflows/containers.yml:38]: Don't we require credentials to push the tag to the registry? Not sure why we're using `||` here.

File: .github/workflows/containers.yml [hunk 2]
```diff
@@ -22,20 +22,31 @@ jobs:
           - dev
           - build
     steps:
+      - name: Check for registry credentials (secrets)
+        id: credentials
+        run: |
+          echo '::set-output name=docker-hub::${{ toJSON(secrets.DOCKER_HUB_USERNAME != null && secrets.DOCKER_HUB_TOKEN != null) }}'
+          echo '::set-output name=ghcr::${{ toJSON(secrets.GHCR_USERNAME != null && secrets.GHCR_TOKEN != null) }}'
       - name: Login to GitHub Container Registry
+        if: ${{ fromJSON(steps.credentials.outputs.ghcr) || startsWith(github.ref, 'refs/tags') }}
```
Comment (petejohanson) [inline .github/workflows/containers.yml:31]: Ditto.

**Review by petejohanson — COMMENTED** [review]
This approach looks much better to me. Happy to merge if your testing is all green.

**Comment by innovaker** [comment]
### Testing

This time, I checked and cleared the registry cache ...
`docker.io/innovaker/zmk-docker-cache:dev`
`docker.io/innovaker/zmk-docker-cache:build`
... before running each scenario.

- all username/token secrets available
  - Candidate: https://github.com/innovaker/zmk-docker/actions/runs/843194978
  - Release: https://github.com/innovaker/zmk-docker/actions/runs/843210247
  - Registry cache created!

- without `DOCKER_HUB_USERNAME`
  - Candidate: https://github.com/innovaker/zmk-docker/actions/runs/843225778
  - Release: https://github.com/innovaker/zmk-docker/actions/runs/843235596
  - Registry cache not created.

- without `DOCKER_HUB_TOKEN`
  - Candidate: https://github.com/innovaker/zmk-docker/actions/runs/843239830
  - Release: https://github.com/innovaker/zmk-docker/actions/runs/843250826
  - Registry cache not created.

- without `GHCR_USERNAME`
  - Candidate: https://github.com/innovaker/zmk-docker/actions/runs/843254916
  - Release: https://github.com/innovaker/zmk-docker/actions/runs/843267433
  - Registry cache created!

- without `GHCR_TOKEN`
  - Candidate: https://github.com/innovaker/zmk-docker/actions/runs/843282722
  - Release: https://github.com/innovaker/zmk-docker/actions/runs/843307954
  - Registry cache created!

- all username/token secrets available (check new tokens)
  - Candidate: https://github.com/innovaker/zmk-docker/actions/runs/843316635
  - Release: https://github.com/innovaker/zmk-docker/actions/runs/843334592
  - Registry cache created!

**Review by petejohanson — APPROVED** [review]

**Pull request merged**